transmit/direct_transmit.go

```diff
@@ -0,0 +1,377 @@
+package transmit
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/tinylib/msgp/msgp"
+	"github.com/vmihailenco/msgpack/v5"
+
+	"github.com/honeycombio/refinery/config"
+	"github.com/honeycombio/refinery/logger"
+	"github.com/honeycombio/refinery/metrics"
+	"github.com/honeycombio/refinery/types"
+)
+
+type transmitKey struct {
+	apiHost string
+	apiKey  string
+	dataset string
+}
+
+// Transmission to the hny API (peer refinery or honeycomb) via messagepack,
+// without involving libhoney. This is designed to be lightweight and sheds
+// some of the ergonomics of libhoney in favor of simplicity and performance.
+// Sending data directly also gives us the flexibility to use custom
+// serialization and potentially send additional metadata to peers.
+type DirectTransmission struct {
+	Config    config.Config   `inject:""`
+	Logger    logger.Logger   `inject:""`
+	Metrics   metrics.Metrics // constructed, not injected
+	Version   string          `inject:"version"`
+	Transport *http.Transport `inject:"upstreamTransport"`
+
+	// Type is peer or upstream, and used only for naming metrics
+	Name string
+
+	// Batching configuration
+	MaxBatchSize int
+	BatchTimeout time.Duration
+
+	transmitQueues map[transmitKey]chan<- *types.Event
+	mutex          sync.RWMutex
+	shutdownWG     sync.WaitGroup
+
+	httpClient *http.Client
+	userAgent  string
+}
+
+func NewDirectTransmission(m metrics.Metrics, name string, maxBatchSize int, batchTimeout time.Duration) *DirectTransmission {
+	return &DirectTransmission{
+		Metrics:        m,
+		Name:           name,
+		MaxBatchSize:   maxBatchSize,
+		BatchTimeout:   batchTimeout,
+		transmitQueues: make(map[transmitKey]chan<- *types.Event),
+	}
+}
+
+func (d *DirectTransmission) Start() error {
+	d.Logger.Debug().Logf("Starting DirectTransmission: %s type", d.Name)
+	d.userAgent = fmt.Sprintf("refinery/%s %s (%s/%s)", d.Version, strings.Replace(runtime.Version(), "go", "go/", 1), runtime.GOOS, runtime.GOARCH)
+	d.httpClient = &http.Client{
+		Transport: d.Transport,
+		Timeout:   10 * time.Second,
+	}
+	return nil
+}
+
+func (d *DirectTransmission) EnqueueEvent(ev *types.Event) {
+	d.Logger.Debug().
+		WithField("request_id", ev.Context.Value(types.RequestIDContextKey{})).
+		WithString("api_host", ev.APIHost).
+		WithString("dataset", ev.Dataset).
+		Logf("transmit sending event")
+
+	// Store enqueue time for queue time metrics
+	ev.EnqueuedUnixMicro = time.Now().UnixMicro()
+
+	// TODO this may end up on the heap, if so, either construct it inline, or make it
+	// a sub-struct of Event (which is already on the heap).
+	key := transmitKey{
```

Comment:
If the amount of traffic per key combination is unevenly distributed, the dataset has large amount of traffic will be capped by a single batch processing goroutine. That seems not ideal

Reply:
Yes, I should have noted in the PR description. This entire queue and dispatch mechanism is a placeholder to allow end-to-end testing, and is NOT viable for production use yet. My plan to to shift to a finite set of workers which are able to share load.